### PR TITLE
#42 HtHead can be broken

### PR DESCRIPTION
**Pull request by SharplEr**
For #42 

**Problem:**
HtHead can miss the separator `\r\n\r\n` when it split by two buffers.
**Solution**
Fix it by using a state machine.

Code a bit complex, I have added a puzzle to fix it.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#56](https://github.com/yegor256/cactoos-http/issues/56) is now in scope, role is `REV`

**Comment by VsSekorin** [comment]
@SharplEr Hi, I'm not REV, but maybe my comment will help you.
- FSM is overkill.
My suggestion:
Input: `1234567890\r\n\r\n0987654321`
n: 6
temp is array for check. temp size = `n + "\r\n\r\n" - 1` = `n + 3` = 9
Read #.1: `123456`. temp: `000123456`
Read #.2: `7890\r\n`. temp: `4567890\r\n`
Read #.4: `\r\n0987`. temp: `0\r\n\r\n0987` - **problem solved**

- `StateMachine` looks like not OOP class (for example, it is not imlemented some interface)

File: src/main/java/org/cactoos/http/io/StateMachine.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+ * Simple immutable finite-state machine for finding a specific pattern
+ * in a sequence of buffers.
+ *
+ * <p>There is thread-safety guarantee.
+ *
+ * @author Alexander Menshikov (sharplermc@gmail.com)
+ * @version $Id$
+ * @since 0.1
+ */
+public final class StateMachine {
```
Comment (llorllale) [inline src/main/java/org/cactoos/http/io/StateMachine.java]: @SharplEr please make this package-private

**Comment by llorllale** [comment]
@SharplEr definitely over-engineered. I left you a comment, see above.

File: src/main/java/org/cactoos/http/io/StateMachine.java (hunk 1 above)
Comment (SharplEr) [reply src/main/java/org/cactoos/http/io/StateMachine.java]: @llorllale Sorry but it follows to a compilation failure. I have put this class in *.http.io package as `BoundedByteBuffer`. I can extract it to *.http and then make it package-private.

**Comment by SharplEr** [comment]
@llorllale 
I have thought about simpler solutions and all of them will not general. FSM we can use in other classes when want to cut buffered input with any separator. Which going to be a common operation in this lib as I see.
We already have class `BoundedByteBuffer` which is support for other classes.
@VsSekorin Your solution needs to copy temp array every time. FSM is a classic and faster solution for such problem.

File: src/test/java/org/cactoos/http/HtHeadTest.java [hunk 2]
```diff
@@ -107,4 +109,138 @@ public void largeText() throws IOException {
...
+                )
+            ).asString(),
+            new MatcherOf<>(
+                str -> str.length() == size * 2 - 4 && str.endsWith("b")
+            )
+        );
+    }
+
+    @Test
+    public void rollbackState() throws IOException {
```
Comment (llorllale) [inline src/test/java/org/cactoos/http/HtHeadTest.java]: @SharplEr what is this test doing?

Comment (SharplEr) [reply src/test/java/org/cactoos/http/HtHeadTest.java]: @llorllale just increases code coverage to an acceptable level.

Comment (llorllale) [reply src/test/java/org/cactoos/http/HtHeadTest.java]: @SharplEr the test must be testing *something*. Let's make its name more descriptive please

**Comment by 0crat** [comment]
This pull request [#56](https://github.com/yegor256/cactoos-http/issues/56) is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/C63314D6Z/1800632460); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @llorllale[/z](https://www.0crat.com/u/llorllale) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by victornoel** [comment]
@SharplEr I'm taking over as reviewer.

I understand your point of view, but I feel that this state machine is both overkill and very specific (it is just to find pattern).

I wonder if the `BoundedByteBuffer` does not provide exactly the functionality you need: to read bytes incrementally but efficiently (without extra copies and stuffs) and be able to check what was read.
Maybe it could be improved to better support your needs.
I think it is currently used only in `SkipInput` which need the same kind of thing: find a pattern.

What do you think?

**Comment by SharplEr** [comment]
@victornoel `BoundedByteBuffer` works extremely slow with extra copies. I have created the issue about it #60 already. With `SkipInput` there are two usages of this class -- not so bad for me. Also even with `BoundedByteBuffer` we still need some state-machine-like code for finding the pattern. I mean even in the @VsSekorin solution we should somehow find the pattern inside the `BoundedByteBuffer`, right? And anyway it will be something like "try the first byte and if match try the second and so on". Which is state machine exactly.

Okay, we can use `BoundedByteBuffer`+`StateMachine` and code in `HtHead.stream()` will become much clear because we will avoid a lot of cases.

I still think in network library which uses HTTP such class can be quite useful.

**Comment by victornoel** [comment]
@SharplEr sorry to hear that, when I introduced `BoundedByteBuffer` it wasn't the case. The whole purpose of this type of buffer is not to make copies and the kind of usage we are having now are exactly why it exists.
@llorllale I think something went wrong when this modification was accepted in #52 honestly…

@SharplEr concerning the behaviour of HtHead, maybe I'm wrong, but isn't the objective to take everything before `\r\n\r\n`? Why a state machine for this simple problem? Or is there something more complex happening here?
With the current behaviour exposed by `BoundedByteBuffer`, you can read byte by byte and check if you found the patter with `equalTo`.
The only open question is how to keep what was read before. Actually I'm surprised there isn't anything for reading a stream up to a point, maybe it is the time to implement that, instead of this `StateMachine`? Or leave a `@todo` to extract it from `HtHead` so that you don't spend too much time on this…

So, should we wait for #60 to be solved and then, depending on how you want to do thing, you either reimplement whatever is in `StateMachine` (which should be renamed, it is too specific for having such a generic name) or you implement it only inside `HtHead`?
Does this seem good to you? Because I can't accept things as they are now, that's too complex, we need some clearer abstractions.

**Comment by llorllale** [comment]
@victornoel 

> I think something went wrong when this modification was accepted in #52 honestly…

Noted. Thanks.

**Comment by 0crat** [comment]
@victornoel[/z](https://www.0crat.com/u/victornoel) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by victornoel** [comment]
@0crat waiting for feedbacks and fixes

**Comment by 0crat** [comment]
> @0crat waiting for feedbacks and fixes ([here](https://github.com/yegor256/cactoos-http/issues/56#issuecomment-402395311))

@victornoel It's a code review job [#56](https://github.com/yegor256/cactoos-http/issues/56), can't put it on hold

**Comment by victornoel** [comment]
@llorllale I think we should reject the PR as it is because the bot is getting impatient (and I spent enough time on it already as a REV).
@SharplEr you should keep your branch and reopen a PR when it is good, depending on how you decide to solve the problem.

**Comment by llorllale** [comment]
@victornoel thanks

**Comment by 0crat** [comment]
@elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#56](https://github.com/yegor256/cactoos-http/issues/56) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

**Comment by elenavolokhova** [comment]
@0crat quality acceptable as far as REV did his job right, but PR was not merged and it can't be good in this case.

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova)

**Pull request closed without merging**